alacritty_config_derive/src/de_struct.rs

```diff
@@ -135,14 +135,18 @@ fn field_deserializer(field_streams: &mut FieldStreams, field: &Field) -> Result
                     config.#ident = serde::Deserialize::deserialize(unused).unwrap_or_default();
                 });
             },
-            "deprecated" => {
-                // Construct deprecation message and append optional attribute override.
-                let mut message = format!("Config warning: {} is deprecated", literal);
+            "deprecated" | "removed" => {
+                // Construct deprecation/removal message and append optional attribute override.
+                let mut verb = "is";
+                if parsed.ident == "removed" {
+                    verb = "was";
+                }
```

Comment:
Pushed an alternative that also avoids the `mut` and does not duplicate the rest of the `format!` call.

Reply:
`message` still needs to be `mut` because it is re-bound right below in the conditional that adds the warning.

Reply:
Working on removing that `mut` as well...

Reply:
Okay, now setting `message` without any `mut`.

Reply:
You just made it worse…

The point is to get clean, readable code which doesn't unnecessarily assign variables just to immediately throw away their values. Your new code is completely unreadable.

Reply:
I'm a little confused. Where am I assigning a variable and throwing it away? The existing code in master did that in order to append the optional attribute text, and I removed it. Your suggestion did not compile because it removed the `mut` on `message` without removing this assignment, so I assumed you would prefer an implementation that did not require any re-assignments of bindings, and that is what the new code does. What am I missing here?

Reply:
Apply this.

```diff
diff --git a/alacritty_config_derive/src/de_struct.rs b/alacritty_config_derive/src/de_struct.rs
index 42787365..6b1e29c0 100644
--- a/alacritty_config_derive/src/de_struct.rs
+++ b/alacritty_config_derive/src/de_struct.rs
@@ -137,16 +137,11 @@ fn field_deserializer(field_streams: &mut FieldStreams, field: &Field) -> Result
             },
             "deprecated" | "removed" => {
                 // Construct deprecation/removal message with optional attribute override.
-                let message = format!(
-                    "Config warning: {} {} {}{}",
-                    literal,
-                    if parsed.ident == "removed" { "was" } else { "is" },
-                    parsed.ident,
-                    parsed
-                        .param
-                        .map(|warning| format!("; {}", warning.value()))
-                        .unwrap_or_default()
-                );
+                let message = match parsed.ident.as_str() {
+                    "removed" => format!("Config warning: {} was removed", literal),
+                    "deprecated" => format!("Config warning: {} is deprecated", literal),
+                    _ => unreachable!(),
+                };
 
                 // Append stream to log deprecation/removal warning.
                 match_assignment_stream.extend(quote! {
```

Reply:
If there are further changes you want here I'm happy to make them of course, but I don't have much information to go on. My best guess is that you would prefer separate `let` bindings for some or all of the arguments to this `format!` call. My approach before was to use idioms that were already present in the code, which included mutable bindings and discarding values. If there is some other source of style I should be following, let me know.

Reply:
@kchibisov this patch doesn't include the extra warning message from the attribute. I also was attempting to avoid unnecessary duplication, since the first attempt which left "removed" and "deprecated" in separate top-level arms of the match, was shot down for being duplicative.

Reply:
Could rename `let message` to `message_prefix` and combine message later. Like `let message = format!("{} {}", message_prefix, the_reset_you_want)`.

Reply:
The problem is then you need to set `message` to `message_prefix` if there is no attribute parameter, and the code is more verbose than before. I honestly think the code as-is (with a `mut` and a conditional assignment) for this is cleaner, but I interpreted from previous comments that this `mut` should be deleted as well. I will make another attempt that is more readable.

Reply:
Please see the latest. It's one additional line relative to `master`, to choose "was" for "removed", and puts the names in the format string so there isn't just a "{} {} {}". I'll leave the existing `mut` and suffix append alone. I think it's out of scope of this change and there isn't an obviously better way to do it, IMO. If this still is distasteful for some reason, I'd suggest to reconsider just using "is removed". It's both grammatically and semantically correct for what the warning needs to convey, and avoids unnecessary string splicing.